Add helper to ease specifying many Variant constraints

eddiejessup opened the pull request:
I haven't tested this beyond my own use-case, and I pretty much made up the `CouldBeAnyOfF` implementation because I'm not using `VariantF` myself, and I haven't rendered the documentation to see how it looks, but I wanted to get something up to see if you think it's a reasonable implementation and interface.

I've done little type-level stuff before, so feedback is welcome!

Commit:
Add helper to ease specifying many Variant constraints

i-am-tom commented on src/Data/Variant.hs:
```diff
@@ -14,6 +14,7 @@
 {-# LANGUAGE StandaloneDeriving     #-}
 {-# LANGUAGE TypeApplications       #-}
 {-# LANGUAGE TypeFamilies           #-}
```
TypeFamilyDependencies implies TypeFamilies, so this line can go :)

i-am-tom commented on src/Data/Variant.hs:
```diff
@@ -276,6 +279,36 @@ class CouldBeF xs x => CouldBe (xs :: [Type]) (x :: Type) where
+  FlattenConstraints  '[] = ()
+  FlattenConstraints (c ': cs) = (c, FlattenConstraints cs)
+
+type family TypeMap (f :: k -> l) (xs :: [k]) = (c :: [l]) where
```
Maybe `ys` or `xs'` is a better name for `c` here, and I'd probably just call it `Map` now we're in a higher namespace :)

i-am-tom commented on src/Data/Variant.hs:
```diff
@@ -276,6 +279,36 @@ class CouldBeF xs x => CouldBe (xs :: [Type]) (x :: Type) where
 instance CouldBeF xs x => CouldBe xs x where
   throw = throwF . Identity
 
+type family FlattenConstraints (cs :: [Constraint]) = (c :: Constraint) | c -> cs where
```
I _think_ the traditional name for this family is `All` or `Every`

i-am-tom commented on src/Data/Variant.hs:
```diff
@@ -276,6 +279,36 @@ class CouldBeF xs x => CouldBe (xs :: [Type]) (x :: Type) where
+  TypeMap f (x ': xs) = f x ': (TypeMap f xs)
+  TypeMap f '[] = '[]
+
+type family SpreadConstraint (f :: k -> Constraint) (xs :: [k]) = (c :: Constraint) where
```
With the sweet machinery you defined above, we can make these definitions much less terrifying and drop `SpreadConstraint` entirely:

```
type e `CouldBeAnyOf`  xs = All (Map (e `CouldBe`) xs)
type e `CouldBeAnyOfF` xs = All (Map (e `CouldBeF`) xs)
```

i-am-tom commented on src/Data/Variant.hs:
```diff
@@ -276,6 +279,36 @@ class CouldBeF xs x => CouldBe (xs :: [Type]) (x :: Type) where
+-- several types a variant may contain in a single type level list, as opposed
+-- to several independent constraints. So, we could replace,
+--
+-- f :: (e `CouldBe` Int, e `CouldBe` Bool, e `CouldBe` Char) => VariantF IO e
```
The comments are actually checked with [`doctest`](http://hackage.haskell.org/package/doctest), so it might be nice to throw in a couple (`>>>`-prefixed) examples of each, just to show that `catch` does indeed still cancel out `throw`!

i-am-tom reviewed (changes requested):
Looks so good! I've left a few comments, but none of them particularly huge. If we can talk about them, I can put this library on Hackage where it can be of use to you? :)

Commit:
Blindly implement PR feedback

- Remove redundant extension
- Rename elements to better match convention
- Simplify constraint-spread implementation
- Add doctest examples

eddiejessup commented:
Thanks for the good feedback! I've acted on the points, let me know how things look now.

i-am-tom reviewed (approved):
Looks awesome, tests pass, docs are great :) Thanks!
